apps/cloud_inference/server_app/source/inference_hip_rocal.cpp

```diff
@@ -0,0 +1,962 @@
+        std::tuple<char*,int> endOfSequenceImage(nullptr,0);
+        queueDeviceTagQ[i]->enqueue(endOfSequenceTag);
+        queueDeviceImageQ[i]->enqueue(endOfSequenceImage);
+    }
+    args->lock();
+    info("workMasterInputQ: terminated for %s [scheduled %d images]", clientName.c_str(), totalInputCount);
+    args->unlock();
+}
+
+void InferenceEngineRocalHip::workDeviceInputCopy(int gpu)
+{
+    args->lock();
+    info("workDeviceInputCopy: GPU#%d started for %s", gpu, clientName.c_str());
+    args->unlock();
+    
+    // create HIP stream
+    hipStream_t stream;
+    hipError_t err;
+    if(hipSuccess != hipStreamCreate(&stream))
+      fatal("workDeviceInputCopy: hipStreamCreate(device_id[%d]) failed (%d)", gpu, err);
+
+    int totalBatchCounter = 0, totalImageCounter = 0;
+    for(bool endOfSequenceReached = false; !endOfSequenceReached; ) {
+        PROFILER_START(inference_server_app, workDeviceInputCopyBatch);
+
+        // get an empty HIP buffer and lock the buffer for writing
+        std::pair<void *, void*> input;
+        queueDeviceInputMemIdle[gpu]->dequeue(input);
+        if(input.first == nullptr) {
+            fatal("workDeviceInputCopy: unexpected nullptr in queueDeviceInputMemIdle[%d] buffer", gpu);
+        }
+        void * mapped_ptr = input.second;
+        if(mapped_ptr == nullptr) {
+            fatal("workDeviceInputCopy: unexpected nullptr in queueDeviceInputMemIdle[%d] mapped_ptr", gpu);
+        }
+        
+        int rem_images = rocalGetRemainingImages(rocalHandle[gpu]);
+        int inputCount = rem_images < batchSize ? rem_images : batchSize;
+
+        for(int i=0; i < inputCount; i++) {
+            // get next item from the input queue and check for end of input
+            std::tuple<char*,int> image;
+            queueDeviceImageQ[gpu]->dequeue(image);
+            int size = std::get<1>(image);
+            if(size == 0) {
+                endOfSequenceReached = true;
+                break;
+            }
+        }
+        
+        high_resolution_clock::time_point t1 = high_resolution_clock::now();
+
+        // decode and resize using rocAL
+        if(rocalRun(rocalHandle[gpu]) != 0) {
+            fatal("workDeviceInputCopy: rocalRun() failed for gpu : [%d]", gpu);
+        }
+        
+        RocalTensorOutputType tensor_output_type = useFp16 ? RocalTensorOutputType::ROCAL_FP16 : RocalTensorOutputType::ROCAL_FP32;
+        RocalTensorLayout tensor_format = RocalTensorLayout::ROCAL_NCHW;
+        
+        if(rocalCopyToOutputTensor(rocalHandle[gpu], mapped_ptr, tensor_format, tensor_output_type, preprocessMpy[0],
+            preprocessMpy[1], preprocessMpy[2], preprocessAdd[0], preprocessAdd[1], preprocessAdd[2], reverseInputChannelOrder) != 0) {
+            fatal("workDeviceInputCopy: rocalCopyToOutputTensor() failed for gpu : [%d]", gpu);
+        }
+
+        mCount+=inputCount;
+        high_resolution_clock::time_point t2 = high_resolution_clock::now();
```

Comment:
rocAL timing has to be called once at the end after all the images are processed. Calling here won't be accurate